### PR TITLE
set GPIO direction (output pin) in uboot

### PR DESCRIPTION
**Pull request by tanchunhau**
Set these pin as output during uboot
(from line 104 - 120)

STATUS_LED_RED_PIN (MXM3_110)
gpio clear gpio9_7

STATUS_LED_GREEN_PIN (MXM3_18)
gpio clear gpio1_20

START_BUTTON_LED (MXM3_120)
gpio clear gpio9_6

UART_2_MUX_CTRL (MXM3_3)
gpio clear gpio3_22

UART_3_MUX_CTRL (MXM3_5)
gpio clear gpio3_23

UART_4_MUX_CTRL (MXM3_7)
gpio clear gpio3_24

PS : For those pin in IO Expander, may need to set in device-tree later.
Please help me check if I miss any, thank you !

Please refer to this :
https://www.toradex.com/community/questions/62318/imx8x-setting-default-gpio-value-in-device-tree.html?childToView=62877#answer-62877

gpio clear [gpio-number]  = set pin to output direction with default value 0 (LOW)

### CODE REVIEW

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+fi
+
+env set bootcmd_args 'run rootfsargs_set && env set bootargs ${defargs} ${rootfsargs} ${setupargs} ${vidargs} ${tdxargs}'
+env set fdt_resize 'fdt addr ${fdt_addr_r} && fdt resize 0x20000'
+env set bootcmd_overlays 'run load_overlays_file && run fdt_resize && run apply_overlays'
+env set bootcmd_boot 'echo "Bootargs: \${bootargs}" && @@KERNEL_BOOTCMD@@ ${kernel_addr_r} - ${fdt_addr_r}'
+env set bootcmd_prepare 'run set_bootcmd_kernel; run set_bootcmd_dtb; run set_load_overlays_file; run set_apply_overlays'
+env set bootcmd_run 'run m4boot; run bootcmd_dtb && run bootcmd_overlays && run bootcmd_args && run bootcmd_kernel && run bootcmd_unzip && run bootcmd_boot; echo "Booting from ${devtype} failed!" && false'
+
+# STATUS_LED_RED_PIN (MXM3_110)
```
Comment (tanchunhau) [inline recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: @NIwkoe , @HatsyRei , @karhin1990 . Please help me verify if I miss any gpio pin. Thank you.

Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: @HatsyRei please help me double check if I key-in "gpio bank & line" number correctly 😃  thanks

Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Looks like you've covered everything. By the way, do we need to care about input pin? Is ALL pins will be in input mode if we didn't explicitly set it?

Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: All pin will be "input" & "active-high" by default.

Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Cool, thanks for the information.

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in [hunk 2]
```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: GPL-2.0+ OR MIT
```
Comment (karhin1990) [inline recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: I would recommend not to overwrite the boot.cmd.in file provided by toradex, reason being we want to be able to pull the latest version from toradex if there's any. This also saves our effort to maintain the content of this boot.cmd.in file.

Is it possible to create a patch file or any extension file such that uboot will also execute it along with the original boot.cmd.in?

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 1 above)
Comment (karhin1990) [inline recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Can you also comment what is the command below doing? E.g.
`# STATUS_LED_RED_PIN (MXM3_110) - Set default direction: OUTPUT; default value: LOW`

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 2 above)
Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: I was thinking this way originally.
after discuss with @HatsyRei , we go with overwrite.
Reason:
If there is a "new file" provided by toradex, we may need to do "diff" and new generate patch file again,
so this case may just similar with overwrite.

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 1 above)
Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: done.

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 2 above)
Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: If they only introduce new changes to the original file without creating a conflict, the patch can still be used right?

Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: the line number need to be same/ dont break.

now we call these gpio set direction before line 122.

run bootcmd_prepare
run bootcmd_run

@HatsyRei correct me if I am wrong.

Comment (HatsyRei) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Yeah, I am primarily concerned with the line numbers in the .patch file. According to what I understand, these define the context allowing do_patch to correctly apply the patch. Seems like Yocto is capable of fuzzy patching, by ignoring some context lines to allow the build to succeed though.

https://www.yoctoproject.org/pipermail/yocto/2018-March/040273.html

If we don't plan on changing our modifications to the boot script regularly, then we can probably go with patching, but if we do, then the extra steps to generate the patch may add a bit more cost to our workflow. These are my personal thoughts after dealing with the IO expander driver patch recently.

Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Hmm, I expect it should works, else a patch file need to be updated whenever a new line is added into the ori file. This seems to defeat the purpose of using patch. Anyways, I don't have a certain as well, will let Yan Huan or comment.

Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: Ah, just saw Yan Huan's response after posting my reply. What Yan Huan said (about the hurdle of updating patch) is actually valid, but I don't expect we need to always regenerate a patch file unless there's a bug in the patch file. If we want to add new functionality to the base file, we will create a separated patch. 

Considering the effort to maintain Toradex code[1] and how often we would regenerate a patch, I would advice to use a patch to set GPIO direction. However, if this "patch approach" takes you more than a day of effort, let's stick with the current approach and create a Todo to fix it later.

[1] By maintaining our own copy of boot.cmd.in, we will never know if the ori file has been updated (can't get latest bug fix/functionality) after we bumped the version of meta-toradex in our manifest file. Even we know, we still need to manually port the new changes to our file. And the beauty of patch file in Yocto is to solve this kind of problem.

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 1 above)
Comment (karhin1990) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: @tanchunhau - if it's not too much trouble perhaps we can help Wilson to handle the GPIO 11 in this PR as well?

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 2 above)
Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: done (changed to patch method)

File: recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in (hunk 1 above)
Comment (tanchunhau) [reply recipes-bsp/u-boot/u-boot-distro-boot/boot.cmd.in]: added GPIO 11 (MXM3_11 , gpio3_17)

**Comment by tanchunhau** [comment]
-change to patch method
-added GPIO 11 (MXM3_11 , gpio3_17)

Please let me know if I still miss any GPIO pin (output)

**Review by karhin1990 — APPROVED** [review]
Thanks for adding this! I didn't really verify the MMX_xyz to the gpioA_B mapping, will count on you to test it out.

**Comment by tanchunhau** [comment]
Drop this PR.
Reason : those pin will be set as output in tGPIOManager initialization

**Pull request closed without merging**